### PR TITLE
feat(remote_wal): implement topic allocation

### PR DESCRIPTION
**Pull request by niebayes**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

- Implemented Kafka topic manager and topic selector to support allocating kafka wal topics.
- Implemented `WalOptionsAllocator` so wal options could be allocated while the metasrv is handling a create table request.

### TODO
- [ ] Integration test to check the topic manager could successfully contact kafka cluster and create topics.

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
Tracing issue for Kafka remote wal: https://github.com/GreptimeTeam/greptimedb/issues/2722

### CODE REVIEW

File: src/common/meta/src/wal/options_allocator.rs [hunk 1]
```diff
@@ -34,37 +34,62 @@ pub enum WalOptionsAllocator {
...
-                    .into_iter()
-                    .map(|topic| WalOptions::Kafka(KafkaWalOptions { topic }))
-                    .collect()
-            }
+            WalOptionsAllocator::Kafka(topic_manager) => topic_manager
+                .select_batch(num_regions)
+                .into_iter()
+                .map(|topic| {
+                    WalOptions::Kafka(KafkaWalOptions {
+                        topic: topic.clone(),
```
Comment (MichaelScofield) [inline src/common/meta/src/wal/options_allocator.rs:73]: Isn't `into_iter` already own the topic?
```suggestion
                        topic,
```

File: src/common/meta/src/wal/kafka/topic_manager.rs [hunk 2]
```diff
@@ -12,40 +12,113 @@
...
             kv_backend,
         }
     }
 
-    /// Tries to initialize the topic pool.
+    /// Tries to initialize the topic manager.
     /// The initializer first tries to restore persisted topics from the kv backend.
-    /// If not enough topics retrieved, the initializer will try to contact the Kafka cluster and request more topics.
+    /// If not enough topics retrieved, the initializer will try to contact the Kafka cluster and request creating more topics.
     pub async fn try_init(&mut self) -> Result<()> {
```
Comment (WenyXu) [inline src/common/meta/src/wal/kafka/topic_manager.rs:63]: We can not invoke the `try_init` on every meta server. Only the leader of the mete server can do this.

File: src/common/meta/src/wal/options_allocator.rs [hunk 3]
```diff
@@ -34,37 +34,62 @@ pub enum WalOptionsAllocator {
...
     }
 }
 
+/// Creates and initializes a wal options allocator.
+pub async fn build_wal_options_allocator(
+    config: &WalConfig,
+    kv_backend: &KvBackendRef,
+) -> Result<WalOptionsAllocator> {
+    let mut allocator = WalOptionsAllocator::new(config.clone(), kv_backend.clone());
+    allocator.try_init().await?;
```
Comment (killme2008) [inline src/common/meta/src/wal/options_allocator.rs:87]: I think the `try_init` must be called outside of this building function. It's better to keep functions orthogonal, one function only tries to do one thing.

Comment (killme2008) [reply src/common/meta/src/wal/options_allocator.rs:87]: Another suggestion: we often use `start` for such initialization procedures.

File: src/common/meta/src/wal/kafka/topic_manager.rs [hunk 4]
```diff
@@ -54,8 +127,60 @@ impl TopicManager {
...
+            .map(|_| self.topic_selector.select(&self.topic_pool))
+            .collect()
+    }
+
+    async fn restore_created_topics(kv_backend: &KvBackendRef) -> Result<Vec<Topic>> {
+        kv_backend
+            .get(CREATED_TOPICS_KEY.as_bytes())
+            .await?
+            .map(|key_value| serde_json::from_slice(&key_value.value).context(DecodeJsonSnafu))
+            .unwrap_or_else(|| Ok(vec![]))
```
Comment (evenyag) [inline src/common/meta/src/wal/kafka/topic_manager.rs:141]: It ignores errors while deserializing the value. I think we should return an error instead.

File: src/common/meta/src/wal/options_allocator.rs [hunk 5]
```diff
@@ -79,13 +104,30 @@ pub fn build_region_wal_options(
...
-    Ok(WalOptionsAllocator::default())
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::kv_backend::memory::MemoryKvBackend;
+
+    // Tests the wal options allocator could successfully allocate raft-engine wal options.
+    // Note: tests for allocator with kafka are integration tests.
+    #[tokio::test]
+    async fn test_allocator_with_raft_engine() {
```
Comment (killme2008) [inline src/common/meta/src/wal/options_allocator.rs:115]: The test is not enough, we don't test kafka allocator at all.

File: src/common/meta/src/wal/kafka/topic_selector.rs [hunk 6]
```diff
@@ -29,23 +31,56 @@ pub enum SelectorType {
 /// Controls topic selection.
 pub(super) trait TopicSelector: Send + Sync {
     /// Selects a topic from the topic pool.
-    fn select(&self, topic_pool: &[Topic]) -> &Topic;
+    fn select<'a>(&'a self, topic_pool: &'a [Topic]) -> &Topic;
 }
```
Comment (v0y4g3r) [inline src/common/meta/src/wal/kafka/topic_selector.rs:35]:
```suggestion]
    fn select<'a>(&self, topic_pool: &'a [Topic]) -> &'a Topic;
```

File: src/common/meta/src/wal/kafka/topic_selector.rs [hunk 7]
```diff
@@ -29,23 +31,56 @@ pub enum SelectorType {
...
 
 impl RoundRobinTopicSelector {
-    /// Creates a new round-robin topic selector.
-    pub(super) fn new() -> Self {
-        todo!()
+    // The cursor in the round-robin selector is not persisted which may break the round-robin strategy cross crashes.
+    // Introducing a shuffling strategy may help mitigate this issue.
+    pub fn with_shuffle() -> Self {
+        let mut this = Self::default();
+        let offset = rand::thread_rng().gen::<usize>() % usize::MAX;
```
Comment (v0y4g3r) [inline src/common/meta/src/wal/kafka/topic_selector.rs:51]: Is this mod operation here necessary?

Comment (evenyag) [inline src/common/meta/src/wal/kafka/topic_selector.rs:51]: I think we can generate the offset from a small range of random numbers.

File: src/common/meta/src/wal/options_allocator.rs (hunk 3 above)
Comment (evenyag) [reply src/common/meta/src/wal/options_allocator.rs:87]: Can we initialize the allocator in `new()`?

File: src/common/meta/src/wal/kafka/topic_selector.rs [hunk 8]
```diff
@@ -29,23 +31,56 @@ pub enum SelectorType {
 /// Controls topic selection.
 pub(super) trait TopicSelector: Send + Sync {
     /// Selects a topic from the topic pool.
-    fn select(&self, topic_pool: &[Topic]) -> &Topic;
+    fn select<'a>(&'a self, topic_pool: &'a [Topic]) -> &Topic;
```
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/common/meta/src/wal/kafka/topic_selector.rs:34]: The `'a` could be dropped safely.

File: src/common/meta/src/error.rs [hunk 9]
```diff
@@ -296,6 +296,37 @@ pub enum Error {
...
+    InvalidNumTopics {
+        num_topics: usize,
+        location: Location,
+    },
+
+    #[snafu(display(
+        "Failed to build an rskafka client, broker endpoints: {:?}",
+        broker_endpoints
+    ))]
+    BuildClient {
```
Comment (fengjiachun) [inline src/common/meta/src/error.rs:310]:
```suggestion
    BuildKafkaClient {
```

File: src/common/meta/src/error.rs [hunk 10]
```diff
@@ -296,6 +296,37 @@ pub enum Error {
...
     },
+
+    #[snafu(display("Invalid number of topics {}", num_topics))]
+    InvalidNumTopics {
+        num_topics: usize,
+        location: Location,
+    },
+
+    #[snafu(display(
+        "Failed to build an rskafka client, broker endpoints: {:?}",
```
Comment (fengjiachun) [inline src/common/meta/src/error.rs:307]:
```suggestion
        "Failed to build a kafka client, broker endpoints: {:?}",
```

File: src/common/meta/src/error.rs [hunk 11]
```diff
@@ -296,6 +296,37 @@ pub enum Error {
...
+        broker_endpoints
+    ))]
+    BuildClient {
+        broker_endpoints: Vec<String>,
+        location: Location,
+        #[snafu(source)]
+        error: rskafka::client::error::Error,
+    },
+
+    #[snafu(display("Failed to build an rskafka controller client"))]
```
Comment (fengjiachun) [inline src/common/meta/src/error.rs:317]:
```suggestion
    #[snafu(display("Failed to build a kafka controller client"))]
```

File: src/common/meta/src/error.rs [hunk 12]
```diff
@@ -296,6 +296,37 @@ pub enum Error {
...
+    BuildCtrlClient {
```
Comment (fengjiachun) [inline src/common/meta/src/error.rs:318]:
```suggestion
    BuildKafkaCtrlClient {
```

File: src/common/meta/src/error.rs [hunk 13]
```diff
@@ -296,6 +296,37 @@ pub enum Error {
...
+        location: Location,
+        #[snafu(source)]
+        error: rskafka::client::error::Error,
+    },
+
+    #[snafu(display("Failed to create a kafka wal topic through rskafka client"))]
```
Comment (fengjiachun) [inline src/common/meta/src/error.rs:324]:
```suggestion
    #[snafu(display("Failed to create a kafka wal topic"))]
```

File: src/common/meta/src/error.rs (hunk 11 above)
Comment (fengjiachun) [inline src/common/meta/src/error.rs:317]:
```suggestion
    #[snafu(display("Failed to build a kafka controller client"))]
```

File: src/common/meta/src/error.rs (hunk 9 above)
Comment (fengjiachun) [reply src/common/meta/src/error.rs:310]: We may have other clients in the future, so the name...

File: src/common/meta/src/error.rs (hunk 11 above)
Comment (fengjiachun) [inline src/common/meta/src/error.rs:317]:
```suggestion
    #[snafu(display("Failed to build a kafka controller client"))]
```

File: src/common/meta/src/error.rs (hunk 10 above)
Comment (fengjiachun) [reply src/common/meta/src/error.rs:307]: I think we will only have one Kafka client, regardless of whether it is rskafka or something else.

File: src/common/meta/src/error.rs (hunk 11 above)
Comment (fengjiachun) [inline src/common/meta/src/error.rs:317]:
```suggestion
    #[snafu(display("Failed to build a kafka controller client"))]
```

File: src/common/meta/src/wal/kafka/topic_manager.rs [hunk 14]
```diff
@@ -12,40 +12,113 @@
...
+    EncodeJsonSnafu, InvalidNumTopicsSnafu, Result,
+};
 use crate::kv_backend::KvBackendRef;
+use crate::rpc::store::PutRequest;
 use crate::wal::kafka::topic::Topic;
 use crate::wal::kafka::topic_selector::{RoundRobinTopicSelector, SelectorType, TopicSelectorRef};
 use crate::wal::kafka::KafkaConfig;
 
+const CREATE_TOPIC_TIMEOUT: i32 = 5_000; // 5,000 ms.
+const CREATED_TOPICS_KEY: &str = "__created_kafka_wal_topics";
```
[the remark this answers is not in the document]
Comment (fengjiachun) [reply src/common/meta/src/wal/kafka/topic_manager.rs:35]: We should let all the keys have a common prefix.

File: src/common/meta/src/wal/kafka/topic_manager.rs (hunk 4 above)
Comment (niebayes) [reply src/common/meta/src/wal/kafka/topic_manager.rs:141]: Changed to the following: 
```rust
    async fn restore_created_topics(kv_backend: &KvBackendRef) -> Result<Vec<Topic>> {
        kv_backend
            .get(CREATED_TOPICS_KEY.as_bytes())
            .await?
            .map_or_else(
                || Ok(vec![]),
                |key_value| serde_json::from_slice(&key_value.value).context(DecodeJsonSnafu),
            )
    }
```

[the remark this answers is not in the document]
Comment (niebayes) [reply src/common/meta/src/wal/kafka/topic_manager.rs:141]: Changed to the following: 
```rust

    async fn restore_created_topics(kv_backend: &KvBackendRef) -> Result<Vec<Topic>> {
        kv_backend
            .get(CREATED_TOPICS_KEY.as_bytes())
            .await?
            .map_or_else(
                || Ok(vec![]),
                |key_value| serde_json::from_slice(&key_value.value).context(DecodeJsonSnafu),
            )
    }
```

File: src/common/meta/src/wal/kafka/topic_selector.rs [hunk 15]
```diff
@@ -29,23 +31,56 @@ pub enum SelectorType {
...
+}
 
 impl RoundRobinTopicSelector {
-    /// Creates a new round-robin topic selector.
-    pub(super) fn new() -> Self {
-        todo!()
+    // The cursor in the round-robin selector is not persisted which may break the round-robin strategy cross crashes.
+    // Introducing a shuffling strategy may help mitigate this issue.
+    pub fn with_shuffle() -> Self {
+        let mut this = Self::default();
```
[the remark this answers is not in the document]
Comment (niebayes) [reply src/common/meta/src/wal/kafka/topic_selector.rs:50]: Replaced `this` with `selector`.

File: src/common/meta/src/wal/kafka/topic_selector.rs (hunk 6 above)
Comment (niebayes) [reply src/common/meta/src/wal/kafka/topic_selector.rs:35]: Updated.

File: src/common/meta/src/wal/kafka/topic_manager.rs [hunk 16]
```diff
@@ -12,40 +12,113 @@
...
+            // Stop reconnecting if the total wait time reaches the deadline.
+            deadline: Some(Duration::from_secs(60 * 5)), // 5 mins.
+        };
+        let client = ClientBuilder::new(broker_endpoints.clone())
+            .backoff_config(backoff_config)
+            .build()
+            .await
+            .context(BuildClientSnafu { broker_endpoints })?
+            .controller_client()
+            .context(BuildCtrlClientSnafu)?;
```
[the remark this answers is not in the document]
Comment (niebayes) [reply src/common/meta/src/wal/kafka/topic_manager.rs:82]: Updated.

File: src/common/meta/src/error.rs (hunk 10 above)
Comment (niebayes) [reply src/common/meta/src/error.rs:307]: Updated.

File: src/common/meta/src/error.rs (hunk 11 above)
Comment (niebayes) [reply src/common/meta/src/error.rs:317]: Updated.

File: src/common/meta/src/error.rs (hunk 12 above)
Comment (niebayes) [reply src/common/meta/src/error.rs:318]: Updated.

File: src/common/meta/src/error.rs (hunk 9 above)
Comment (niebayes) [reply src/common/meta/src/error.rs:310]: Updated.

File: src/common/meta/src/error.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (niebayes) [reply src/common/meta/src/error.rs:307]: Changed to `build a kafka client`.

File: src/common/meta/src/error.rs (hunk 13 above)
Comment (niebayes) [reply src/common/meta/src/error.rs:324]: Updated.

File: src/common/meta/src/error.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (niebayes) [reply src/common/meta/src/error.rs:317]: Changed to `build a kafka controller client`.

File: src/common/meta/src/wal/kafka/topic_manager.rs [hunk 17]
```diff
@@ -12,40 +12,113 @@
...
+    BuildClientSnafu, BuildCtrlClientSnafu, CreateKafkaWalTopicSnafu, DecodeJsonSnafu,
+    EncodeJsonSnafu, InvalidNumTopicsSnafu, Result,
+};
 use crate::kv_backend::KvBackendRef;
+use crate::rpc::store::PutRequest;
 use crate::wal::kafka::topic::Topic;
 use crate::wal::kafka::topic_selector::{RoundRobinTopicSelector, SelectorType, TopicSelectorRef};
 use crate::wal::kafka::KafkaConfig;
 
+const CREATE_TOPIC_TIMEOUT: i32 = 5_000; // 5,000 ms.
```
[the remark this answers is not in the document]
Comment (niebayes) [reply src/common/meta/src/wal/kafka/topic_manager.rs:34]: It's now set to 30 seconds by default and it's exposed as a config.

File: src/common/meta/src/wal/kafka/topic_manager.rs [hunk 18]
```diff
@@ -12,40 +12,113 @@
...
+                debug!("Tries to create topic {}", topic);
+                Some(client.create_topic(
+                    topic,
+                    self.config.num_partitions,
+                    self.config.replication_factor,
+                    CREATE_TOPIC_TIMEOUT,
+                ))
+            })
+            .collect::<Vec<_>>();
+        // TODO(niebayes): Determine how rskafka handles an already-exist topic. Check if an error would be raised.
```
[the remark this answers is not in the document]
Comment (niebayes) [reply src/common/meta/src/wal/kafka/topic_manager.rs:113]: The following PR will update the CI to start a kafka cluster like what we did on etcd. That PR will also contain enough tests.

File: src/common/meta/src/wal/kafka/topic_manager.rs [hunk 19]
```diff
@@ -54,8 +127,60 @@ impl TopicManager {
...
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::kv_backend::memory::MemoryKvBackend;
+
+    // Tests that topics can be successfully persisted into the kv backend and can be successfully restored from the kv backend.
+    #[tokio::test]
```
[the remark this answers is not in the document]
Comment (niebayes) [reply src/common/meta/src/wal/kafka/topic_manager.rs:163]: Good suggestion. But it's the rskafka client needs to be mocked or the kafka server needs to be mocked?

File: src/common/meta/src/wal/options_allocator.rs (hunk 3 above)
Comment (niebayes) [reply src/common/meta/src/wal/options_allocator.rs:87]: I prefer to do the initialization in `new` or `try_new`. But some comments suggest to not mix `new` and `start`. The advantage is that the method name implies that a heavy operation may be executed by the `start` method.

File: src/common/meta/src/wal/options_allocator.rs (hunk 5 above)
Comment (niebayes) [reply src/common/meta/src/wal/options_allocator.rs:115]: Will add more integration tests in the following PR.

File: src/common/meta/src/wal/kafka/topic_manager.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (niebayes) [reply src/common/meta/src/wal/kafka/topic_manager.rs:34]: It's now in the KafkaConfig.

File: src/common/meta/src/error.rs (hunk 11 above)
Comment (niebayes) [reply src/common/meta/src/error.rs:317]: Updated.

File: src/common/meta/src/wal/kafka/topic_manager.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (niebayes) [reply src/common/meta/src/wal/kafka/topic_manager.rs:35]: Renamed to `"__created_wal_topics/kafka/"`

[the remark this answers is not in the document]
Comment (niebayes) [reply src/common/meta/src/wal/kafka/topic_manager.rs:35]: If there're another topic system, their topics would be `"__created_wal_topics/other_topic_system"`

File: src/common/meta/src/wal/kafka/topic_manager.rs [hunk 20]
```diff
@@ -12,40 +12,113 @@
...
+
+        // Builds an rskafka controller client for creating topics if necessary.
+        let broker_endpoints = self.config.broker_endpoints.clone();
+        let backoff_config = BackoffConfig {
+            init_backoff: Duration::from_millis(500),
+            max_backoff: Duration::from_secs(10),
+            base: 2.0,
+            // Stop reconnecting if the total wait time reaches the deadline.
+            deadline: Some(Duration::from_secs(60 * 5)), // 5 mins.
+        };
```
[the remark this answers is not in the document]
Comment (niebayes) [reply src/common/meta/src/wal/kafka/topic_manager.rs:75]: They're integrated into the `KafkaConfig` and exposed to users.

[the remark this answers is not in the document]
Comment (niebayes) [reply src/common/meta/src/wal/kafka/topic_manager.rs:75]: They're now integrated into the `KafkaConfig` and configurable by users.

File: src/common/meta/src/wal/kafka/topic_manager.rs [hunk 21]
```diff
@@ -12,40 +12,113 @@
...
+                ))
+            })
+            .collect::<Vec<_>>();
+        // TODO(niebayes): Determine how rskafka handles an already-exist topic. Check if an error would be raised.
+        futures::future::try_join_all(tasks)
+            .await
+            .context(CreateKafkaWalTopicSnafu)?;
+
+        // Persists created topics.
+        Self::persist_created_topics(&topics, &self.kv_backend).await?;
```
[the remark this answers is not in the document]
Comment (niebayes) [reply src/common/meta/src/wal/kafka/topic_manager.rs:119]: Used a `num_created_topics` to record how many topics are newly created. If it's greater than 0, the persistence operation will be executed.

File: src/common/meta/src/wal/options_allocator.rs (hunk 3 above)
Comment (niebayes) [reply src/common/meta/src/wal/options_allocator.rs:87]: The `try_init` method is renamed to `start` and it's called in `MetaStateHandler::on_become_leader`.

File: src/common/meta/src/wal/kafka/topic_manager.rs (hunk 2 above)
Comment (niebayes) [reply src/common/meta/src/wal/kafka/topic_manager.rs:63]: The try_init method is renamed to start and it's called in MetaStateHandler::on_become_leader.

File: src/common/meta/src/wal/kafka/topic_manager.rs [hunk 22]
```diff
@@ -12,40 +12,113 @@
...
+                Some(client.create_topic(
+                    topic,
+                    self.config.num_partitions,
+                    self.config.replication_factor,
+                    CREATE_TOPIC_TIMEOUT,
+                ))
+            })
+            .collect::<Vec<_>>();
+        // TODO(niebayes): Determine how rskafka handles an already-exist topic. Check if an error would be raised.
+        futures::future::try_join_all(tasks)
```
[the remark this answers is not in the document]
Comment (niebayes) [reply src/common/meta/src/wal/kafka/topic_manager.rs:114]: We now first check if there're any topics should be created. If there're some, the kafka client is created then.

File: src/common/meta/src/wal/kafka/topic_selector.rs [hunk 23]
```diff
@@ -29,23 +31,56 @@ pub enum SelectorType {
...
+        let offset = rand::thread_rng().gen::<usize>() % usize::MAX;
+        // It's ok when an overflow happens since `fetch_add` automatically wraps around.
+        this.cursor.fetch_add(offset, Ordering::Relaxed);
+        this
```
[the remark this answers is not in the document]
Comment (niebayes) [reply src/common/meta/src/wal/kafka/topic_selector.rs:54]: Simplified.

File: src/common/meta/src/wal/kafka/topic_selector.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (niebayes) [reply src/common/meta/src/wal/kafka/topic_selector.rs:34]: The lifetime seems necessary. Or there will have an error: 
```rust
error: lifetime may not live long enough
  --> src/common/meta/src/wal/kafka/topic_selector.rs:60:9
   |
58 |     fn select(&self, topic_pool: &[Topic]) -> &Topic {
   |               -                  - let's call the lifetime of this reference `'1`
   |               |
   |               let's call the lifetime of this reference `'2`
59 |         let which = self.cursor.fetch_add(1, Ordering::Relaxed) % topic_pool.len();
60 |         &topic_pool[which]
   |         ^^^^^^^^^^^^^^^^^^ method was supposed to return data with lifetime `'2` but it is returning data with lifetime `'1`
   |
help: consider introducing a named lifetime parameter and update trait if needed
   |
58 |     fn select<'a>(&'a self, topic_pool: &'a [Topic]) -> &Topic {
   |              ++++  ++                    ++


```

File: src/common/meta/src/wal/kafka/topic_selector.rs (hunk 7 above)
Comment (niebayes) [reply src/common/meta/src/wal/kafka/topic_selector.rs:51]: The offset is now randomly generated from a small range and hence the mod op is removed.

File: src/common/meta/src/wal/options_allocator.rs (hunk 1 above)
Comment (niebayes) [reply src/common/meta/src/wal/options_allocator.rs:73]: `select_batch` returns a vector of topic references.

File: src/common/meta/src/wal/kafka/topic_selector.rs [hunk 24]
```diff
@@ -29,23 +31,56 @@ pub enum SelectorType {
...
 }
 
 impl TopicSelector for RoundRobinTopicSelector {
-    fn select(&self, topic_pool: &[Topic]) -> &Topic {
-        todo!()
+    fn select<'a>(&'a self, topic_pool: &'a [Topic]) -> &Topic {
+        // Safety: the caller ensures the topic pool is not empty and hence the modulo operation is safe.
+        let which = self.cursor.fetch_add(1, Ordering::Relaxed) % topic_pool.len();
+        // Safety: the modulo operation ensures the index operation is safe.
+        topic_pool.get(which).unwrap()
```
[the remark this answers is not in the document]
Comment (niebayes) [reply src/common/meta/src/wal/kafka/topic_selector.rs:63]: Updated.

File: src/common/meta/src/wal/kafka/topic_selector.rs (hunk 7 above)
Comment (niebayes) [reply src/common/meta/src/wal/kafka/topic_selector.rs:51]: The offset is now generated from range 0..64.

File: src/common/meta/src/wal/kafka/topic_selector.rs [hunk 25]
```diff
@@ -29,23 +31,56 @@ pub enum SelectorType {
...
+        this.cursor.fetch_add(offset, Ordering::Relaxed);
+        this
     }
 }
 
 impl TopicSelector for RoundRobinTopicSelector {
-    fn select(&self, topic_pool: &[Topic]) -> &Topic {
-        todo!()
+    fn select<'a>(&'a self, topic_pool: &'a [Topic]) -> &Topic {
+        // Safety: the caller ensures the topic pool is not empty and hence the modulo operation is safe.
```
[the remark this answers is not in the document]
Comment (niebayes) [reply src/common/meta/src/wal/kafka/topic_selector.rs:60]: `select` now checks that the input topic pool is not empty. And the return type is changed to `Result<&Topic>`.

File: src/common/meta/src/wal/kafka/topic_selector.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (niebayes) [reply src/common/meta/src/wal/kafka/topic_selector.rs:34]: The `select` signature is updated to:
```rust
    fn select<'a>(&self, topic_pool: &'a [Topic]) -> Result<&'a Topic>;
```

File: src/common/meta/src/wal/kafka/topic_manager.rs (hunk 21 above)
[the remark this answers is not in the document]
Comment (niebayes) [reply src/common/meta/src/wal/kafka/topic_manager.rs:119]: There has a `to_be_created` vector to record which topics are going to be created. If this vector is not empty, the persistence operation will be executed.

File: src/common/meta/src/wal/kafka/topic_manager.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (niebayes) [reply src/common/meta/src/wal/kafka/topic_manager.rs:181]: Will add a mock kafka client and/or mock kafka client when necessary during adding integrations tests.

**Review by killme2008 — APPROVED** [review]
LGTM

File: config/metasrv.example.toml [hunk 26]
```diff
@@ -61,11 +61,21 @@ provider = "raft_engine"
 # - "round_robin" (default)
 # selector_type = "round_robin"
 # A Kafka topic is constructed by concatenating `topic_name_prefix` and `topic_id`.
-# topic_name_prefix = "greptimedb_kafka_wal"
+# topic_name_prefix = "greptimedb_wal_kafka"
 # Number of partitions per topic.
 # num_partitions = 1
 # Expected number of replicas of each partition.
 # replication_factor = 3
+# Above which a topic creation operation will be cancelled.
```
Comment (WenyXu) [inline config/metasrv.example.toml:69]: Let's use more concise words.

Nit(Non-blocking):  
```suggestion
# The timeout of topic creation
```

**Review by MichaelScofield — APPROVED** [review]

File: src/meta-srv/src/metasrv.rs [hunk 27]
```diff
@@ -196,6 +198,11 @@ impl MetaStateHandler {
         if let Err(e) = self.procedure_manager.start().await {
             error!(e; "Failed to start procedure manager");
         }
+
+        if let Err(e) = self.wal_options_allocator.start().await {
+            error!(e; "Failed to start wal options allocator");
+        }
```
Comment (WenyXu) [inline src/meta-srv/src/metasrv.rs:204]: Nit(Non-blocking): 
We should ensure the start will only be executed once in the life cycle. We can improve this in the future.

**Pull request merged**